src/vw/src/main/scala/VowpalWabbitBase.scala

```diff
@@ -0,0 +1,410 @@
+// Copyright (C) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License. See LICENSE in project root for information.
+
+package com.microsoft.ml.spark
+
+import java.io.{BufferedReader, BufferedWriter, InputStreamReader, OutputStreamWriter}
+import java.net.{InetAddress, ServerSocket, Socket}
+import java.util.UUID
+import java.util.concurrent.Executors
+
+import org.apache.spark.{ClusterUtil, TaskContext}
+import org.apache.spark.ml.linalg.{DenseVector, SparseVector}
+import org.apache.spark.ml.param._
+import org.apache.spark.ml.util.DefaultParamsWritable
+import org.apache.spark.sql.functions.{col, struct, udf}
+import org.apache.spark.sql.types.StructType
+import org.apache.spark.sql.{DataFrame, Dataset, Encoders, Row}
+import org.apache.spark.internal._
+import org.vowpalwabbit.spark.prediction.ScalarPrediction
+import org.vowpalwabbit.spark.{ClusterSpanningTree, VowpalWabbitExample, VowpalWabbitMurmur, VowpalWabbitNative}
+
+import scala.math.min
+import scala.concurrent.{ExecutionContext, Future}
+
+case class NamespaceInfo (hash: Int, featureGroup: Char, colIdx: Int)
+
+object VWUtil {
+  /**
+    * Implementation of C# using(...) { } pattern.
+    */
+  def autoClose[A <: AutoCloseable,B](closeable: A)(fun: (A) => B): B = {
+    try {
```

Comment:
we have a similar function in StreamUtilities called `using` perhaps re-use that if you need

Reply:
thanks! didn't know. done.